util_for_test.go

```diff
@@ -0,0 +1,329 @@
+package surrealdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	rawslog "log/slog"
+	"net"
+	"os/exec"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/surrealdb/surrealdb.go/pkg/conn/gorilla"
+	"github.com/surrealdb/surrealdb.go/pkg/logger/slog"
+)
+
+var (
+	testHasSurrealCLI     bool
+	checkSurrealCLIOnce   sync.Once
+	DelayAfterServerStart time.Duration = 300 * time.Millisecond
+	DelayBeforeServerExit time.Duration = 300 * time.Millisecond
+)
+
+func Test_NewTestSurrealDB_Simple(t *testing.T) {
+	// Simple illustration of how NewTestSurrealDB can be used.
+
+	// Tests can run in parallel.
+	t.Parallel()
+
+	// NewTestSurrealDB returns endpoint (random open port used), DB instance
+	// (which is wrapped in DBForTest to provide extra methods), and a func to
+	// shutdown the SurrealDB server.
+	endpoint, db, close := NewTestSurrealDB(t)
+	defer close()
+	_ = endpoint // endpoint could be useful for some low level testing.
+
+	// DBForTest.Prepare is to run any SurrealQL for test prep. This would fail
+	// early if the provided string results in an error from SurrealDB server.
+	db.Prepare(t, "CREATE user:x SET name = 'x'")
+
+	// More interactions can simply use db instance from here on.
+}
+
+func Test_TestSurrealDBParallelSimple(t *testing.T) {
+	t.Parallel()
+	instanceCount := 20
+	for i := 0; i < instanceCount; i++ {
+		t.Run(fmt.Sprintf("parallel__%d", i), func(t *testing.T) {
+			t.Parallel()
+
+			_, db, close := NewTestSurrealDB(t)
+			defer close()
+
+			// Ensure that CREATE statement does not run against the same
+			// database (which would fail).
+			db.Prepare(t, "CREATE user:x SET name = 'x';")
+		})
+	}
+}
+
+func Test_TestSurrealDBParalleSelect(t *testing.T) {
```

Comment:
This test does `t.Parallel()` in the subtests, and that was the main point. In a way this is the main test to ensure that `surreal start` happens in parallel with different ports, so that there will be no conflicting interactions for SurrealQL to fail.